llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp

```diff
@@ -8131,15 +8131,73 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
 void SelectionDAGBuilder::visitConstrainedFPIntrinsic(
     const ConstrainedFPIntrinsic &FPI) {
   SDLoc sdl = getCurSDLoc();
+  const TargetLowering &TLI = DAG.getTargetLoweringInfo();
+  EVT VT = TLI.getValueType(DAG.getDataLayout(), FPI.getType());
+  fp::ExceptionBehavior EB = *FPI.getExceptionBehavior();
+  std::optional<RoundingMode> RM = FPI.getRoundingMode();
+
+  SDNodeFlags Flags;
+  if (EB == fp::ExceptionBehavior::ebIgnore)
+    Flags.setNoFPExcept(true);
+
+  if (auto *FPOp = dyn_cast<FPMathOperator>(&FPI))
+    Flags.copyFMF(*FPOp);
+
+  bool UseStaticRounding = EB == fp::ExceptionBehavior::ebIgnore && RM &&
+                           *RM != RoundingMode::Dynamic &&
```

Comment:
Isn't the contract for the constrained intrinsics, that the global rounding mode has already been changed? What do we gain by using the static rounding mode instructions?

If the global rounding mode hasn't already been changed, that means that whoever created the constrained intrinsics needs to know what operations support a static rounding mode in the target.

Reply:
Yes. These would be useful if they were connected intrinsics with a fixed rounding mode (like we have fptrunc.round)

Reply:
Right... if we want floating-point ops with an explicitly specified rounding mode, we need IR support.  And "STRICT_" forms of the SelectionDAG ops.  If you want to go this route, it needs an RFC.

(If you just want the nodes for SelectionDAG-internal optimizations, you can skip this, though.)

Reply:
The forthcoming C standard introduces static rounding mode in addition to the dynamic one (set by `fesetround`): https://www.iso-9899.info/n3047.html#7.6.2 . This rounding is set by pragma STDC FENV_ROUND and may be different from the dynamic rounding. This model well fits architecture of targets, where rounding mode can be encoded in an instruction. RISCV is such in the LLVM tree, but there are other processors (maybe most AI cores) that support this feature. There is a demand to support static rounding mode in LLVM.

Constrained intrinsics are not the best representation for static rounding mode in IR, because they have side effect, but the can be used as a starting point. In future new ways to express associated rounding mode can appear, may be as proposed in https://discourse.llvm.org/t/thought-on-strictfp-support/71453.

Reply:
But that's not what you are implementing here. You are implementing operations that do read the dynamic mode into ones that don't. We would need a separate set of operations corresponding to cases with a static mode. The metadata is not a request to use a specific mode, they only indicate what the mode is.

> Constrained intrinsics are not the best representation for static rounding mode in IR

These aren't a representation of it at all, it's conceptually different.

Reply:
What is wrong with using constrained intrinsic for operations with static rounding?

LangRef defines these intrinsics (https://llvm.org/docs/LangRef.html#constrained-floating-point-intrinsics):

> These intrinsics are used to provide special handling of floating-point operations when specific rounding mode or floating-point exception behavior is required. 

This is just what is required for static rounding operations. The bad thing is the side effect, which is present on every constrained intrinsic. It however does not break the semantics of static rounding, it only prevents some optimization.

Constrained intrinsics are used here as per their desigs, no?

Reply:
> What is wrong with using constrained intrinsic for operations with static rounding?

It's not really what they are for. This isn't wrong, but why would using the fixed mode instructions be better than using instructions that do read the mode? The constrained argument only tells you the dynamic mode will be a certain value. Using fixed mode instructions does not change the expected behavior. Is there some reason using the fixed mode instructions is an optimization (I'd assume not)?

> Constrained intrinsics are used here as per their desigs, no?

Not really. What you're describing is more like https://llvm.org/docs/LangRef.html#llvm-fptrunc-round-intrinsic, where the rounding mode argument provides the behavior to use. The backend must manage mode switches as necessary, or use fixed rounding mode instructions like these ones you are defining.